### PR TITLE
Allow mods to temporary prevent DEs from completing

### LINKED ISSUE #596: Allow mods to temporary prevent DEs from completing

### PR DESCRIPTION
**Pull request by Xymanek**
Closes #596 

~~[Better compare](https://github.com/X2CommunityCore/X2WOTCCommunityHighlander/compare/1eca2b885e808dcae6a5e8445b6a3112680cda7a...460080496db565157cac47bbf291ebdd041b8a94)~~ seems no longer necessary after the rebase

### CODE REVIEW

**Comment by Xymanek** [comment]
> This seems to be an unusual approach for the highlander, which normally uses events to allow mods to influence vanilla behaviour

This is generally done out of necessity, because the behaviour is in either

* Native gamestate classes, where we can't change the data layout
* Non-gamestate classes, where an event/dlc info hook is the only way to make things non-static (non-config)

> why not use an event 

Because in this case we aren't bound by the 2 limitations above

> check whether any mods want to keep the dark event active?

That's not what this PR does though. It controls whether the DE can transition from preparing to active (NOT from active to expired)

**Comment by MrNiceUK** [comment]
Agree with calling it "Activation" not "Completion". On the Hungarian notation point, for whateverreason it's far from uncommon in the XCom 2 UC source to use the `b` for bool flags even when Hungarian isn't used for other properties/variables. Speculation: I think it's because you can't past bools as out parameters in UC, so in some areas flags are stored as 0/1 ints, which often also have the `i` prefix, where it has to pass through out parameters (see that a lot in the various step/peek out related methods for example). So may have just become normalized by Firaxis devs to note if a flag is a bool or an int...
(aside, I can't seem to find what to click on to actually see the detail of  your requested change?)

**Comment by Xymanek** [comment]
I agree that the naming is confusing, but I went with FXS's terms that you can see elsewhere in code. For example, 2 lines later: `if(DarkEventState.CanComplete())`

On the point of `Temporary`: I specifically included it to make it obvious that the flag isn't meant for permanent changes (like forever disabling the DE)

**Comment by pledbrook** [comment]
Then it should be `bTemporarilyPreventCompletion`, but I still don't think that really helps with clarity. To be honest, I think it would be clearer to have "BlockCompletion()" and "UnblockCompletion()" functions instead.

**Comment by Xymanek** [comment]
Do you mean making the flag protected/private and have only the functions update it?

**Comment by pledbrook** [comment]
Yes. It makes it clearer that it's something one activates and then deactivates. It's not clear how to use the variable otherwise.

**Comment by Xymanek** [comment]
I'm not certain that adding mutator funcs for a boolean member is a good solution.... Would extensive commentary around the variable be a good alternative?

**Comment by pledbrook** [comment]
> I'm not certain that adding mutator funcs for a boolean member is a good solution

Why not? It will also allow the opportunity to accept a `NewGameState` parameter, which if not provided, will result in the mutator function setting up a new game state change container for the modification.

**Comment by Musashi1584** [comment]
Open change requests still valid? @Xymanek

**Comment by Xymanek** [comment]
Docs fail sees unrelated

File: X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_DarkEvent.uc [hunk 1]
```diff
@@ -23,6 +23,15 @@ var TDateTime						   EndDateTime;
 var float							   TimeRemaining;
 var StrategyCost					   RevealCost;
 
+// Issue #596
+/// HL-Docs: feature:TemporarilyBlockDarkEventActivation; issue:597; tags:strategy,compat
```
Comment (robojumper) [inline X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_DarkEvent.uc:27]:
```suggestion
/// HL-Docs: feature:TemporarilyBlockDarkEventActivation; issue:597; tags:strategy,compatibility
```

If this has the compatibility tag, a section describing the potential compatibility issue and workarounds/fixes is required, see https://x2communitycore.github.io/X2WOTCCommunityHighlander/compatibility/

Alternatively, remove the tag

Comment (Xymanek) [reply X2WOTCCommunityHighlander/Src/XComGame/Classes/XComGameState_DarkEvent.uc]: Oof, I didn't even notice that was there. Removed - it shouldn't be

**Comment by robojumper** [comment]
> 
> 
> Docs fail sees unrelated

pushed 83463250227150a0ad6369b5fd5627c94095a9b5 to #899 to catch this sort of issue

**Pull request merged**